File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/AnnouncementLogic.kt
package com.lextures.android.core.lms

import com.lextures.android.core.navigation.MobilePlatformFeatures

enum class AnnouncementAudience {
    WholeCourse,
    Section,
}

enum class BroadcastComposeType(val wire: String) {
    Announcement("announcement"),
    Emergency("emergency"),
}

object AnnouncementLogic {
    const val ORG_BROADCAST_MANAGE_PERMISSION = "tenant:org:roles:manage"
    const val GLOBAL_ADMIN_PERMISSION = "global:app:rbac:manage"

    fun canComposeCourseAnnouncement(course: CourseSummary): Boolean =
        course.viewerIsStaff && course.isFeedEnabled

    fun canComposeBroadcast(permissions: List<String>, features: MobilePlatformFeatures): Boolean {
        if (!features.ffBroadcasts) return false
        return ORG_BROADCAST_MANAGE_PERMISSION in permissions || GLOBAL_ADMIN_PERMISSION in permissions
    }

    fun resolveOrgId(courses: List<CourseSummary>): String? =
        courses.mapNotNull { it.orgId?.trim() }.firstOrNull { it.isNotEmpty() }

    fun announcementsChannelId(channels: List<FeedChannel>): String? =
        channels.firstOrNull { it.name.equals("announcements", ignoreCase = true) }?.id

    fun formatAnnouncementBody(
        title: String,
        body: String,
        sectionName: String?,
        mentionsEveryone: Boolean,
    ): String {
        val trimmedTitle = title.trim()
        val trimmedBody = body.trim()
        var text = "**$trimmedTitle**\n\n$trimmedBody"
        val section = sectionName?.trim().orEmpty()
        if (section.isNotEmpty()) {
            text += "\n\n_($section)_"
        }
        if (mentionsEveryone) {
            text += "\n\n@everyone"
        }
        return text
    }

    fun audienceLabel(
        course: CourseSummary,
        audience: AnnouncementAudience,
        sectionName: String?,
    ): String = when (audience) {
        AnnouncementAudience.WholeCourse -> course.displayTitle
        AnnouncementAudience.Section -> {
            val name = sectionName?.trim().orEmpty()
            if (name.isEmpty()) course.displayTitle else "${course.displayTitle} · $name"
        }
    }

    fun canSubmitCourseAnnouncement(title: String, body: String): Boolean =
        title.trim().isNotEmpty() && body.trim().isNotEmpty()

    fun canSubmitBroadcast(subject: String, body: String): Boolean =
        subject.trim().isNotEmpty() && body.trim().isNotEmpty()
}

File: clients/ios/Lextures/Core/LMS/AnnouncementLogic.swift
import Foundation

enum AnnouncementAudience: String, CaseIterable, Hashable {
    case wholeCourse
    case section
}

enum BroadcastComposeType: String, CaseIterable, Hashable {
    case announcement
    case emergency
}

enum AnnouncementLogic {
    static let orgBroadcastManagePermission = "tenant:org:roles:manage"
    static let globalAdminPermission = "global:app:rbac:manage"

    static func canComposeCourseAnnouncement(course: CourseSummary) -> Bool {
        course.viewerIsStaff && course.isFeedEnabled
    }

    static func canComposeBroadcast(permissions: [String], features: MobilePlatformFeatures) -> Bool {
        guard features.ffBroadcasts else { return false }
        return permissions.contains(orgBroadcastManagePermission)
            || permissions.contains(globalAdminPermission)
    }

    static func resolveOrgId(courses: [CourseSummary]) -> String? {
        courses.compactMap(\.orgId).first { !$0.isEmpty }
    }

    static func announcementsChannelId(channels: [FeedChannel]) -> String? {
        channels.first { $0.name.lowercased() == "announcements" }?.id
    }

    static func formatAnnouncementBody(
        title: String,
        body: String,
        sectionName: String?,
        mentionsEveryone: Bool
    ) -> String {
        let trimmedTitle = title.trimmingCharacters(in: .whitespacesAndNewlines)
        let trimmedBody = body.trimmingCharacters(in: .whitespacesAndNewlines)
        var text = "**\(trimmedTitle)**\n\n\(trimmedBody)"
        if let sectionName, !sectionName.isEmpty {
            text += "\n\n_(\(sectionName))_"
        }
        if mentionsEveryone {
            text += "\n\n@everyone"
        }
        return text
    }

    static func audienceLabel(
        course: CourseSummary,
        audience: AnnouncementAudience,
        sectionName: String?
    ) -> String {
        switch audience {
        case .wholeCourse:
            return course.displayTitle
        case .section:
            let name = sectionName?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            return name.isEmpty ? course.displayTitle : "\(course.displayTitle) · \(name)"
        }
    }

    static func broadcastReachLabel() -> String {
        L.text("mobile.broadcast.reach.org")
    }

    static func canSubmitCourseAnnouncement(title: String, body: String) -> Bool {
        !title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            && !body.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    static func canSubmitBroadcast(subject: String, body: String) -> Bool {
        !subject.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            && !body.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }
}